File: LC/[0253] Meeting Rooms II - sweep line.cpp
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        map<int, int> Map; //events - {timestamp, diff}
        for(auto interval: intervals){
            Map[interval[0]] += 1;
            Map[interval[1]] -= 1;
        }

        int sum=0;
        int ret=0;
        for(auto & [ts, diff]: Map){
            sum+=diff;
            ret = max(ret,sum);
        }

        return ret;
    }
};

File: LC/[0253] Meeting Rooms II - sweep line.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Map<Integer, Integer> map = new TreeMap<>();
        for(var interval: intervals){
            map.put(interval[0], map.getOrDefault(interval[0], 0)+1);
            map.put(interval[1], map.getOrDefault(interval[1], 0)-1);
        }
        
        int sum=0;
        int mx=0;
        for(var diff: map.values()){
            sum+=diff;
            mx = Math.max(mx, sum);
        }
        return mx;
    }
}